From bd29f6c411e8adfbcbc03296ad559d9355a5ed29 Mon Sep 17 00:00:00 2001
From: Gurpreet Ghai <gghai@codeaurora.org>
Date: Wed, 8 May 2019 17:08:17 +0530
Subject: [PATCH] BT: Aptx Adaptive R2 feature changes

- Aptx Adaptive codec operation in TWS+ mode
- Processing of next gen ALS app inputs

CRs-Fixed: 2454981

Change-Id: Ibea98d59b9e35121adc4b8ca3f8cdd20b34e0081
---
 bta/av/bta_av_aact.cc    | 32 +++++++++++++++++++++++++++++---
 bta/av/bta_av_api.cc     | 15 +++++++++++++++
 bta/av/bta_av_int.h      | 10 ++++++++++
 bta/av/bta_av_main.cc    |  3 +++
 bta/include/bta_av_api.h | 14 ++++++++++++++
 btif/co/bta_av_co.cc     |  5 ++++-
 btif/src/btif_av.cc      |  6 +++---
 stack/avdt/avdt_ccb.cc   |  2 +-
 8 files changed, 79 insertions(+), 8 deletions(-)

diff --git a/bta/av/bta_av_aact.cc b/bta/av/bta_av_aact.cc
index f3598c2..bc14b8b 100644
--- a/bta/av/bta_av_aact.cc
+++ b/bta/av/bta_av_aact.cc
@@ -130,7 +130,7 @@ static void bta_av_vendor_offload_select_codec(tBTA_AV_SCB* p_scb);
 
 void bta_av_vendor_offload_check_stop_start(tBTA_AV_SCB* p_scb);
 void update_sub_band_info(uint8_t **param, int *param_len, uint8_t id, uint16_t data);
-void update_sub_band_info(uint8_t **param, int *param_len, uint8_t id, uint8_t *data, uint8_t len);
+void update_sub_band_info(uint8_t **param, int *param_len, uint8_t id, uint8_t *data, uint8_t size);
 void enc_mode_change_callback(tBTM_VSC_CMPL *param);
 
 /* state machine states */
@@ -2716,6 +2716,24 @@ void bta_av_update_enc_mode(tBTA_AV_DATA* p_data) {
   *num_sub_band += 1;
 
    BTM_VendorSpecificCommand(HCI_VSQC_CONTROLLER_A2DP_OPCODE, param_len,
+                                 param, NULL);
+}
+void bta_av_update_aptx_data(tBTA_AV_DATA* p_data) {
+  uint8_t param[48];
+  uint8_t *p_param;
+  uint8_t *num_sub_band;
+  int param_len = 0;
+  uint8_t subband_id = p_data->aptx_data.type;
+  uint8_t subband_data = p_data->aptx_data.data;
+  memset(param, 0, 48);
+  p_param = param;
+  *p_param++ = VS_QHCI_ENCODER_MODE_CHANGE;
+  param_len++;
+  num_sub_band = p_param++;
+  param_len++;
+  update_sub_band_info(&p_param, &param_len, subband_id, &subband_data, 1);
+  *num_sub_band += 1;
+   BTM_VendorSpecificCommand(HCI_VSQC_CONTROLLER_A2DP_OPCODE, param_len,
                                  param, enc_mode_change_callback);
 }
 
@@ -2733,9 +2751,17 @@ void update_sub_band_info(uint8_t **param, int *p_param_len, uint8_t id, uint16_
   *param = p_param;
 }
 
-void update_sub_band_info(uint8_t **param, int *p_param_len, uint8_t id, uint8_t *data, uint8_t len)
+void update_sub_band_info(uint8_t **param, int *p_param_len, uint8_t id, uint8_t *data, uint8_t size)
 {
-  /* For future use */
+  uint8_t *p_param = *param;
+  *p_param++ = BTA_AV_ENCODER_DATA_ID;
+  *p_param_len += 1;
+  *p_param++ = 2;
+  *p_param_len += 1;
+  *p_param++ = id;
+  *p_param++ = *data;
+  *p_param_len += 2;
+  *param = p_param;
 }
 
 void enc_mode_change_callback(tBTM_VSC_CMPL *param)
diff --git a/bta/av/bta_av_api.cc b/bta/av/bta_av_api.cc
index 661de1d..6414cef 100644
--- a/bta/av/bta_av_api.cc
+++ b/bta/av/bta_av_api.cc
@@ -443,6 +443,21 @@ void BTA_AvUpdateEncoderMode(uint16_t enc_mode) {
   bta_sys_sendmsg(p_buf);
 }
 
+void BTA_AvUpdateAptxData(uint16_t data) {
+  tBTA_AV_APTX_DATA* p_buf =
+      (tBTA_AV_APTX_DATA*)osi_malloc(sizeof(tBTA_AV_APTX_DATA));
+  uint16_t BATTERY_INFO_MASK = 0X0F;
+  uint16_t ULL_MODE_MASK = 0x6000;
+  if(data & BATTERY_INFO_MASK) {
+    p_buf->type = 4;
+    p_buf->data = data;
+  } else if(data & ULL_MODE_MASK) {
+    p_buf->type = 3;
+    p_buf->data = 1;
+  }
+  p_buf->hdr.event = BTA_AV_UPDATE_APTX_DATA_EVT;
+  bta_sys_sendmsg(p_buf);
+}
 
 /*******************************************************************************
  *
diff --git a/bta/av/bta_av_int.h b/bta/av/bta_av_int.h
index 26c0220..b6b749a 100644
--- a/bta/av/bta_av_int.h
+++ b/bta/av/bta_av_int.h
@@ -146,6 +146,7 @@ enum {
   BTA_AV_ENABLE_MULTICAST_EVT, /* Event for enable and disable multicast */
   BTA_AV_RC_COLLISSION_DETECTED_EVT,
   BTA_AV_UPDATE_ENCODER_MODE_EVT,
+  BTA_AV_UPDATE_APTX_DATA_EVT,
 #if (TWS_ENABLED == TRUE)
   BTA_AV_SET_EARBUD_ROLE_EVT, /* Set TWS earbud role */
   BTA_AV_SET_TWS_DEVICE_EVT, /* Update TWS state */
@@ -183,6 +184,8 @@ enum {
 /* Info ID from updating aptX Adaptive Encoder mode */
 #define BTA_AV_ENCODER_MODE_CHANGE_ID 5
 
+#define BTA_AV_ENCODER_DATA_ID 0x0E
+
 /* maximum number of SEPS in stream discovery results */
 #define BTA_AV_NUM_SEPS 32
 
@@ -400,6 +403,11 @@ typedef struct {
   uint16_t enc_mode;
 } tBTA_AV_ENC_MODE;
 
+typedef struct {
+  BT_HDR hdr;
+  uint16_t type;
+  uint16_t data;
+} tBTA_AV_APTX_DATA;
 
 /* data type for BTA_AV_CI_SETCONFIG_OK_EVT and BTA_AV_CI_SETCONFIG_FAIL_EVT */
 typedef struct {
@@ -532,6 +540,7 @@ union tBTA_AV_DATA {
   tBTA_AV_ENABLE_MULTICAST multicast_state;
   tBTA_AV_MAX_CLIENT max_av_clients;
   tBTA_AV_ENC_MODE encoder_mode;
+  tBTA_AV_APTX_DATA aptx_data;
 #if (TWS_ENABLED == TRUE)
   tBTA_AV_TWS_SET_EARBUD_ROLE tws_set_earbud_role;
   tBTA_AV_SET_TWS_DEVICE tws_set_device;
@@ -841,6 +850,7 @@ extern void bta_av_conn_chg(tBTA_AV_DATA* p_data);
 extern void bta_av_dereg_comp(tBTA_AV_DATA* p_data);
 extern void bta_av_rc_collission_detected(tBTA_AV_DATA *p_data);
 extern void bta_av_update_enc_mode(tBTA_AV_DATA* p_data);
+extern void bta_av_update_aptx_data(tBTA_AV_DATA* p_data);
 extern void bta_av_active_browse(tBTA_AV_DATA *p_data);
 
 /* sm action functions */
diff --git a/bta/av/bta_av_main.cc b/bta/av/bta_av_main.cc
index 7fdaaf1..dcd7445 100644
--- a/bta/av/bta_av_main.cc
+++ b/bta/av/bta_av_main.cc
@@ -227,6 +227,7 @@ const tBTA_AV_NSM_ACT bta_av_nsm_act[] = {
     bta_av_api_enable_multicast,    /* BTA_AV_ENABLE_MULTICAST_EVT */
     bta_av_rc_collission_detected, /* BTA_AV_RC_COLLISSION_DETECTED_EVT */
     bta_av_update_enc_mode, /* BTA_AV_UPDATE_ENCODER_MODE_EVT */
+    bta_av_update_aptx_data,     /* BTA_AV_UPDATE_APTX_DATA_EVT */
 #if (TWS_ENABLED == TRUE)
     bta_av_api_set_tws_earbud_role, /* BTA_AV_SET_EARBUD_ROLE_EVT */
     bta_av_api_set_is_tws_device, /* BTA_AV_SET_TWS_DEVICE_EVT */
@@ -1695,6 +1696,8 @@ const char* bta_av_evt_code(uint16_t evt_code) {
       return "MULTICAST_ENABLE";
     case BTA_AV_UPDATE_ENCODER_MODE_EVT:
       return "UPDATE_ENCODER_MODE";
+    case BTA_AV_UPDATE_APTX_DATA_EVT:
+      return "UPDATE_APTX_DATA";
     default:
       return "unknown";
   }
diff --git a/bta/include/bta_av_api.h b/bta/include/bta_av_api.h
index cc5836a..593e149 100644
--- a/bta/include/bta_av_api.h
+++ b/bta/include/bta_av_api.h
@@ -744,6 +744,20 @@ void BTA_AvUpdateEncoderMode(uint16_t enc_mode);
 
 /*******************************************************************************
  *
+ * Function          BTA_AvUpdateAptxData
+ *
+ * Description      Update extended Aptx Data to Soc via
+ *                          Vendor Specific Command.
+ *                          Sends Battery level, scan mode and
+ *                          ULL mode info to SoC
+ *
+ * Returns            void
+ *
+ ******************************************************************************/
+void BTA_AvUpdateAptxData(uint16_t data);
+
+
+/*******************************************************************************
  * Function         BTA_AvProtectReq
  *
  * Description      Send a content protection request.  This function can only
diff --git a/btif/co/bta_av_co.cc b/btif/co/bta_av_co.cc
index 828e3d5..67aa5e4 100644
--- a/btif/co/bta_av_co.cc
+++ b/btif/co/bta_av_co.cc
@@ -1405,12 +1405,15 @@ bool bta_av_co_set_codec_user_config(
         APPL_TRACE_DEBUG("%s: Updating Encoder Mode to: %x", __func__, encoder_mode);
         BTA_AvUpdateEncoderMode(encoder_mode);
       }
+      if(codec_user_config.codec_specific_4 > 0) {
+        APPL_TRACE_DEBUG("%s: Updating Extended Aptx Data: %x", __func__, codec_user_config.codec_specific_4);
+        BTA_AvUpdateAptxData(codec_user_config.codec_specific_4);
+      }
 
       const uint16_t BT_SCAN_CONTROL_MASK = 0xB000;
       const uint16_t BLE_SCAN_ON = 0x9000; //ORing of HQ(0x1000) and BTcontrol(0x8000)
       const uint16_t BLE_SCAN_OFF = 0xA000;//ORing of LL(0x2000) and BTcontrol(0x8000)
       uint16_t blescanonoroff = codec_user_config.codec_specific_4 & BT_SCAN_CONTROL_MASK;
-
       if (blescanonoroff == BLE_SCAN_OFF) {
         APPL_TRACE_DEBUG("%s: Disabling BLE Scanning", __func__);
         btif_gatt_get_interface()->scanner->Scan(false);
diff --git a/btif/src/btif_av.cc b/btif/src/btif_av.cc
index 6e96760..693d49c 100644
--- a/btif/src/btif_av.cc
+++ b/btif/src/btif_av.cc
@@ -4112,13 +4112,13 @@ static bt_status_t codec_config_src(const RawAddress& bd_addr,
   //RawAddress *bda = &bda;
   //check if current device is TWS and then return failure with SHO support
 #if (TWS_ENABLED == TRUE)
-  int index = btif_av_idx_by_bdaddr(const_cast<RawAddress*>(&bd_addr));
-  if (index < btif_max_av_clients && btif_av_cb[index].tws_device) {
+  //int index = btif_av_idx_by_bdaddr(const_cast<RawAddress*>(&bd_addr));
+  /*if (index < btif_max_av_clients && btif_av_cb[index].tws_device) {
   //if (btif_av_is_tws_connected()) {
     BTIF_TRACE_DEBUG("%s:TWSP device connected, config change not allowed",__func__);
     return BT_STATUS_FAIL;
   //}
-  }
+  } */
 #endif
   btif_av_codec_config_req_t codec_req;
   isDevUiReq = false;
diff --git a/stack/avdt/avdt_ccb.cc b/stack/avdt/avdt_ccb.cc
index 87e2fd5..9c67e9a 100644
--- a/stack/avdt/avdt_ccb.cc
+++ b/stack/avdt/avdt_ccb.cc
@@ -384,7 +384,7 @@ void avdt_ccb_event(tAVDT_CCB* p_ccb, uint8_t event, tAVDT_CCB_EVT* p_data) {
   /* execute action functions */
   for (i = 0; i < AVDT_CCB_ACTIONS; i++) {
     action = state_table[event][i];
-    AVDT_TRACE_DEBUG("%s: event=%s state=%s action=%d", __func__,
+    AVDT_TRACE_DEBUG("%s: event=%d=%s state=%s action=%d", __func__, event,
                      avdt_ccb_evt_str[event], avdt_ccb_st_str[p_ccb->state],
                      action);
     if (action != AVDT_CCB_IGNORE) {
-- 
2.7.4

